clippy_lints/src/unnecessary_indexing.rs

```diff
@@ -0,0 +1,163 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::eq_expr_value;
+use clippy_utils::source::snippet;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir::{ExprKind, Local, StmtKind, UnOp};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::{sym, Span};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for the use of `seq.is_empty()` in an if-conditional where `seq` is a slice, array, or Vec,
+    /// and in which the first element of the sequence is indexed.
+    ///
+    /// ### Why is this bad?
+    /// This code is unnecessarily complicated and can instead be simplified to the use of an
+    /// if..let expression which accessed the first element of the sequence.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let a: &[i32] = &[1];
+    /// if !a.is_empty() {
+    ///    let b = a[0];
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let a: &[i32] = &[1];
+    /// if let Some(b) = a.first() {
+    ///
+    /// }
+    /// ```
+    #[clippy::version = "1.78.0"]
+    pub UNNECESSARY_INDEXING,
+    style,
+    "unnecessary use of `seq.is_empty()` in a conditional when if..let is more appropriate"
+}
+
+declare_lint_pass!(UnnecessaryIndexing => [UNNECESSARY_INDEXING]);
+
+impl LateLintPass<'_> for UnnecessaryIndexing {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &'_ rustc_hir::Expr<'_>) {
+        if let Some(if_expr) = clippy_utils::higher::If::hir(expr)
+            // check for negation
+            && let ExprKind::Unary(op, unary_inner) = if_expr.cond.kind
+            && UnOp::Not == op
+            // check for call of is_empty
+            && let ExprKind::MethodCall(method, conditional_receiver, _, _) = unary_inner.kind
+            && let expr_ty = cx.typeck_results().expr_ty(conditional_receiver).peel_refs()
+            && (expr_ty.is_array_slice() || expr_ty.is_array() || is_type_diagnostic_item(cx, expr_ty, sym::Vec))
+            && method.ident.as_str() == "is_empty"
+            && let ExprKind::Block(block, _) = if_expr.then.kind
+        {
+            // checked if conditional is calling `is_empty` on a sequence, now check if the first
+            // statement inside the 'if' statement does unchecked get of first element
+
+            // if calling a function on the indexing - local
+            if let Some(first) = block.stmts.first()
```

Comment:
suggestion: we're matching on a couple kinds of expressions here

however what we really should do is use a visitor to find all indexing expressions in the block

explicit_counter_loop does this, for example https://github.com/rust-lang/rust-clippy/blob/3a149111ba2ad470c43d58cc1879f85209112b8e/clippy_lints/src/loops/explicit_counter_loop.rs

Reply:
(probably worth asking in the [clippy Zulip channel](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy) for help with this)